test_conformance/common/vulkan_wrapper/vulkan_utility.cpp

```diff
@@ -236,30 +237,119 @@ getSupportedVulkanExternalMemoryHandleTypeList()
 }
 
 const std::vector<VulkanExternalSemaphoreHandleType>
-getSupportedVulkanExternalSemaphoreHandleTypeList()
+getSupportedVulkanExternalSemaphoreHandleTypeList(const VulkanDevice &vkDevice)
 {
-    std::vector<VulkanExternalSemaphoreHandleType>
-        externalSemaphoreHandleTypeList;
-
+    typedef struct
+    {
+        const char *extension_name;
+        VkExternalSemaphoreHandleTypeFlagBits vk_type;
+        VulkanExternalSemaphoreHandleType enum_type;
+    } VkSemaphoreHandleMap;
+
+    // Add all known handle types, use Vulkan queries to determine what is
+    // supported.
+    std::vector<VkSemaphoreHandleMap> all_known_handle_types;
+    all_known_handle_types.push_back(
+        { "VK_KHR_external_semaphore_fd",
+          VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_FD_BIT_KHR,
+          VULKAN_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_FD });
+    all_known_handle_types.push_back(
+        { "VK_KHR_external_semaphore_fd",
+          VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_SYNC_FD_BIT_KHR,
+          VULKAN_EXTERNAL_SEMAPHORE_HANDLE_TYPE_SYNC_FD });
+    all_known_handle_types.push_back(
+        { "VK_KHR_external_semaphore_win32",
+          VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_WIN32_KMT_BIT_KHR,
+          VULKAN_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_WIN32_KMT });
+    // An exception to the rule about letting Vulkan sort out support.  Only
+    // allow NT macro on Windows 8 and later
 #if _WIN32
     if (IsWindows8OrGreater())
     {
-        externalSemaphoreHandleTypeList.push_back(
-            VULKAN_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_WIN32_NT);
+        all_known_handle_types.push_back(
+            { "VK_KHR_external_semaphore_win32",
+              VK_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_WIN32_BIT_KHR,
+              VULKAN_EXTERNAL_SEMAPHORE_HANDLE_TYPE_OPAQUE_WIN32_NT });
```

Comment:
Is there any reason why we push opaque_fd, sync_fd and win32 KMT unconditionally, but need to push win32 NT conditionally? If we are anyway going to check against supported handle types later and this is merely a reference for all known handle types, don't see a need to check for Windows version condition here.

Reply:
The original structure of the code used conditional compilation, I tried to maintain the original structure, as I do not understand or have the ability to test windows handles.  For example, we still have this code in the test from the original PR:
`
const std::vector<VulkanExternalMemoryHandleType>
getSupportedVulkanExternalMemoryHandleTypeList()
{
    std::vector<VulkanExternalMemoryHandleType> externalMemoryHandleTypeList;

#if _WIN32
    if (IsWindows8OrGreater())
    {
        externalMemoryHandleTypeList.push_back(
            VULKAN_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_NT);
    }
    externalMemoryHandleTypeList.push_back(
        VULKAN_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_KMT);
#else
    externalMemoryHandleTypeList.push_back(
        VULKAN_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_FD);
#endif

    return externalMemoryHandleTypeList;
}
`

I was unclear on how to query Vulkan for the appropriate handle type.  For example,  the NT enum is a bitwise or of Vulkan flags:

`
VULKAN_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_NT_KMT =
        VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_BIT_KHR
        | VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_KMT_BIT_KHR
`

I am not sure how to interpret this, in the context of querying support from Vulkan for semaphore or memory handles.

Can you advise?

Reply:
Since the test was iterating over all handle types pushed here, it made sense to push them conditionally here earlier so that we don't accidentally test for unsupported handle type based on OS version. Now that we anyway cross check this against supported handle types and use this just as reference of all known handle types, I think we can clean this up and push the types unconditionally.